File: src/org/simbrain/util/piccolo/Tile.java
package org.simbrain.util.piccolo;

import com.thoughtworks.xstream.annotations.XStreamAlias;
import com.thoughtworks.xstream.annotations.XStreamAsAttribute;
import com.thoughtworks.xstream.annotations.XStreamConverter;
import com.thoughtworks.xstream.annotations.XStreamImplicit;
import com.thoughtworks.xstream.converters.extended.NamedMapConverter;
import org.simbrain.util.UserParameter;
import org.simbrain.util.propertyeditor2.EditableObject;

import java.util.HashMap;
import java.util.Map;

/**
 * Simbrain wrapper for a {@link TileMap} tile.
 */
@XStreamAlias("tile")
public class Tile  implements EditableObject {

    // Ideal structure for Odor World version / notes for future javadoc comments
    //
    // TileType (= tmx tile id).
    //  Based on tmx ids, integer.
    //  Display label is "Tile Type ID".
    //  This determines what image is used
    //   Associated with custom properties.
    //   We can add a custom property called SimbrainType, and if filled, use it to set SimbrainType
    //
    // SimbrainTileType (Grass, Water, etc). Initially just use a string.
    //   Maybe later a list of strings if multiple type systems

    /**
     * The local tile ID within its tileset.
     */
    @XStreamAsAttribute
    private int id;

    /**
     * Custom properties defined in tmx.
     */
    @XStreamConverter(
            value = NamedMapConverter.class,
            strings = {"property", "name", "value"},
            types = {String.class, String.class},
            booleans = {true, true}
    )
    private HashMap<String, String> properties;

    /**
     * Label for this tile
     */
    @UserParameter(label = "Label", description = "label", order = 1)
    private transient String label;

    /**
     * Type of this tile. Multiple tiles can be associated with the same type.
     */
    @UserParameter(label = "Type", description = "type", order = 2)
    private transient String type;

    /**
     * Standard method call made to objects after they are deserialized. See:
     * http://java.sun.com/developer/JDCTechTips/2002/tt0205.html#tip2
     * http://xstream.codehaus.org/faq.html
     *
     * @return Initialized object.
     */
    private Object readResolve() {
        label = properties.containsKey("label") ? properties.get("label") : "Tile" + id;
        type = properties.containsKey("type") ?  properties.get("type:") : "" + id;
        return this;
    }

    /**
     * Construct a tile with a given local tileset id.
     * @param id the local tileset id of this tile
     */
    public Tile(int id) {
        this.id = id;
        this.label = "Tile " + id;
        this.type = "Type " + id;
    }

    public int getId() {
        return id;
    }
}

File: src/main/kotlin/org/simbrain/util/piccolo/Tile.kt
package org.simbrain.util.piccolo

import com.thoughtworks.xstream.annotations.XStreamAlias
import com.thoughtworks.xstream.annotations.XStreamAsAttribute
import com.thoughtworks.xstream.annotations.XStreamConverter
import com.thoughtworks.xstream.converters.extended.NamedMapConverter
import org.simbrain.util.UserParameter
import org.simbrain.util.point
import org.simbrain.util.propertyeditor.EditableObject
import org.simbrain.world.odorworld.entities.Bounded

/**
 * Simbrain wrapper for a [TileMap] tile.
 */
@XStreamAlias("tile")
class Tile(@XStreamAsAttribute @UserParameter(label = "ID", order = 0, displayOnly = true) val id: Int) : EditableObject {

    /**
     * Type of this tile. This way multiple tiles can be associated with the same type.
     */
    @Transient
    @UserParameter(label = "Type", description = "type", displayOnly = true, order = 20)
    var type: String = "Type $id"

    /**
     * Human readable name associated to some tile ids. It's up to the person making the tmx file to provide these.
     * Provides a more readable way to set tiles in scripts.
     */
    @Transient
    @UserParameter(label = "Tile label", description = "Name for tile", displayOnly = true, order = 30)
    var label: String? = null

    /**
     * Custom properties defined in tmx.
     */
    @XStreamConverter(
        value = NamedMapConverter::class,
        strings = ["property", "name", "value"],
        types = [String::class, String::class],
        booleans = [true, true]
    )
    val properties = HashMap<String, String>()

    /**
     * See {@link org.simbrain.workspace.serialization.WorkspaceComponentDeserializer}
     */
    private fun readResolve(): Any {
        type = properties["type"] ?: "$id"
        label = properties["label"]
        return this
    }

    override val name = "$id"
}

/**
 * A temporary data structure to associate a tile with location information during collision detection.
 */
context(TileMap)
class TileInstance(val tile: Tile, val gridCoordinate: GridCoordinate): Bounded {
    val pixelCoordinate = gridCoordinate.toPixelCoordinate()
    override val x by pixelCoordinate::x
    override val y by pixelCoordinate::y
    override val location = point(x, y)
    override val width = tileWidth.toDouble()
    override val height = tileHeight.toDouble()
}